### PR TITLE
Automatic data skipping using generated columns

### PR DESCRIPTION
**Pull request by allisonport-db**
## Overview

This pull request introduces automatic data skipping using generated columns, resolving #880. For a partition column with a generation expression, we automatically generate a partition filter from any data filters on its generating column, when possible. This can be turned on/off using the new `DeltaSQLConf.GENERATED_COLUMN_PARTITION_FILTER_OPTIMIZION_ENABLED` with key `"generatedColumn.partitionFilterOptimization.enabled"`.

## Implementation details

- `trait OptimizablePartitionExpression` - rules to convert a data filter to a partition filter for a specific generation expression (of a partition column)

- `GeneratedColumn.getOptimizablePartitionExpressions(schema, partitionSchema)` - uses the table's schema and partition schema to create an `OptimizablePartitionExpression` for any partition column generated from a data column

- `GeneratedColumn.generatePartitionFilters(...)` - generates partition filters from provided data filters if possible

- `PrepareDeltaScan` - expanded to generate partition filters from data filters if possible, and apply them to the logical plan

## Testing

`OptimizeGeneratedColumnSuite` includes unit tests for correct partition filter generation, and detecting conflicting transactions based on read files.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala
Comment (Kimahriman) [inline core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala:35]: Do these need to be part of the same rule? It seems weird to group them together as mostly unrelated

File: core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala
Comment (Kimahriman) [inline core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala:54]: This check seems to have gotten lost, so everything is now turning into a PinnedTahoeFileIndex regardless of if there's an active transaction?

File: core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala:294]: is this the right way to get spark session? cant the session be explicitly passed to this method?

**Comment by akhil393** [comment]
@allisonport-db  It does not look like you are evaluating a mod operation, can you include the mod operation as one of the generated column expression? Thanks.

**Comment by Kimahriman** [comment]
Nested value extraction would be a nice to have too. We have a case where we a bunch of tables in a common format, and some of them are big enough to add another partition layer, but the value we want is in a nested field. If we used a generated column to make the partition value, we could get the partition filtering by filtering the nested value. Obviously it's a slight overhead of just having the same value in the data as well, but should be very well compressed.

File: core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala:35]: I agree that it's a little weird (and could use some renaming). But it helps ensure we use the optimized filters to record read files in an optimistic transaction. However, #966 is removing this file anyways, and we've decided to refactor all of this into [`PrepareDeltaScan`](https://github.com/delta-io/delta/blob/503881673da016ba29b5ca79374819a199a2d613/core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala).

File: core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala
Comment (allisonport-db) [reply core/src/main/scala/org/apache/spark/sql/delta/ActiveOptimisticTransactionRule.scala:54]: Good catch. This should be resolved when moving the code to [PrepareDeltaScan](https://github.com/delta-io/delta/blob/503881673da016ba29b5ca79374819a199a2d613/core/src/main/scala/org/apache/spark/sql/delta/stats/PrepareDeltaScan.scala).

**Comment by allisonport-db** [comment]
> Nested value extraction would be a nice to have too. We have a case where we a bunch of tables in a common format, and some of them are big enough to add another partition layer, but the value we want is in a nested field. If we used a generated column to make the partition value, we could get the partition filtering by filtering the nested value. Obviously it's a slight overhead of just having the same value in the data as well, but should be very well compressed.

I think it's not necessary to include in this PR but could be added on later. Feel free to open an issue suggesting this.

**Comment by allisonport-db** [comment]
> @allisonport-db It does not look like you are evaluating a mod operation, can you include the mod operation as one of the generated column expression? Thanks.

Mod operations are not super useful since they could only support equality filters.

**Review by zsxwing — APPROVED** [review]
LGTM. Given ActiveOptimisticTransactionRule is going to be removed entirely soon, we can leave ActiveOptimisticTransactionRule as it is.

**Comment by zsxwing** [comment]
@allisonport-db could you resolve the conflicts before merging?

**Comment by allisonport-db** [comment]
Could I get a second review @zsxwing? I updated this to work with #966

**Review by zsxwing — APPROVED** [review]
Awesome. Thanks for resolving the huge conflicts. LGTM!

**Comment by Kimahriman** [comment]
I started working on trying add direct nested value extraction, and it looks like nested values aren't supported at all right now, even for supported expressions, is that expected? For example `date -> CAST(nested.timestamp as DATE)` wouldn't currently be supported

**Pull request closed without merging**
